services/appflowy-collaborate/src/indexer/document_indexer.rs

```diff
@@ -141,3 +112,211 @@ impl Indexer for DocumentIndexer {
     }))
   }
 }
+
+fn create_embedding_params(
+  object_id: String,
+  content: String,
+  collab_type: CollabType,
+  max_content_len: usize,
+) -> Result<Vec<AFCollabEmbeddingParams>, AppError> {
+  let mut result = Vec::with_capacity(1 + content.len() / max_content_len);
+
+  let mut start = 0;
+  let mut current_len = 0;
+  let graphemes: Vec<&str> = content.graphemes(true).collect();
```

Comment:
In 90% of the cases you won't ever need to split the string at all, so allocating array of graphemes seems wasteful.